src/realm/util/file_mapper.cpp

```diff
@@ -1,379 +1,499 @@
+{
+    return (size + page_size() - 1) & ~(page_size() - 1);
+}
+
+void* mmap(FileDesc fd, size_t size, File::AccessMode access, size_t offset, const char* encryption_key,
+           EncryptedFileMapping*& mapping)
+{
+    if (encryption_key) {
+        size = round_up_to_page_size(size);
+        void* addr = mmap_anon(size);
+        mapping = add_mapping(addr, size, fd, offset, access, encryption_key);
+        return addr;
+    }
+    else {
+        mapping = nullptr;
+        return mmap(fd, size, access, offset, nullptr);
+    }
+}
+
+#endif // enable encryption
+
+
+void* mmap(FileDesc fd, size_t size, File::AccessMode access, size_t offset, const char* encryption_key)
+{
+#if REALM_ENABLE_ENCRYPTION
+    if (encryption_key) {
+        size = round_up_to_page_size(size);
+        void* addr = mmap_anon(size);
+        add_mapping(addr, size, fd, offset, access, encryption_key);
+        return addr;
+    }
+    else
+#else
+    REALM_ASSERT(!encryption_key);
+#endif
+    {
+
+#ifndef _WIN32        
+        int prot = PROT_READ;
+        switch (access) {
+        case File::access_ReadWrite:
+            prot |= PROT_WRITE;
+            break;
+        case File::access_ReadOnly:
+            break;
+        }
+
+        void* addr = ::mmap(nullptr, size, prot, MAP_SHARED, fd, offset);
+        if (addr != MAP_FAILED)
+            return addr;
+
+        int err = errno; // Eliminate any risk of clobbering
+        if (is_mmap_memory_error(err)) {
+            throw AddressSpaceExhausted(get_errno_msg("mmap() failed: ", err) + " size: " + util::to_string(size) +
+                " offset: " + util::to_string(offset));
+        }
+
+        throw std::runtime_error(get_errno_msg("mmap() failed: ", err) + "size: " + util::to_string(size) + "offset: " +
+            util::to_string(offset));
+
+#else
+        // FIXME: Is there anything that we must do on Windows to honor map_NoSync?
+
+        DWORD protect = PAGE_READONLY;
+        DWORD desired_access = FILE_MAP_READ;
+        switch (access) {
+        case File::access_ReadOnly:
+            break;
+        case File::access_ReadWrite:
+            protect = PAGE_READWRITE;
+            desired_access = FILE_MAP_WRITE;
+            break;
+        }
+        LARGE_INTEGER large_int;
+        if (int_cast_with_overflow_detect(offset + size, large_int.QuadPart))
+            throw std::runtime_error("Map size is too large");
+        HANDLE map_handle = CreateFileMappingFromApp(fd, 0, protect, offset + size, nullptr);
+        if (!map_handle)
+            throw std::runtime_error("CreateFileMapping() failed");
+
+        if (int_cast_with_overflow_detect(offset, large_int.QuadPart))
+            throw std::runtime_error("Map offset is too large");
+
+        SIZE_T _size = size;
+        void* addr = MapViewOfFileFromApp(map_handle, desired_access, offset, _size);
+        BOOL r = CloseHandle(map_handle);
+        REALM_ASSERT_RELEASE(r);
+        if (!addr)

```

Comment:
Should specifically check for whatever the appropriate error code is for `AddressSpaceExhausted` since that produces a different error in some of the bindings.